lib/linter/source-code-traverser.js

```diff
@@ -218,39 +196,115 @@ class NodeEventGenerator {
 						selectorsByNodeType[selectorsByNodeTypeIndex],
 					) < 0)
 			) {
-				this.applySelector(
+				this.#applySelector(
 					node,
+					ancestry,
 					anyTypeSelectors[anyTypeSelectorsIndex++],
 				);
 			} else {
 				// otherwise apply the node type selector
-				this.applySelector(
+				this.#applySelector(
 					node,
+					ancestry,
 					selectorsByNodeType[selectorsByNodeTypeIndex++],
 				);
 			}
 		}
 	}
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
 
+/**
+ * Traverses source code and ensures that visitor methods are called when
+ * entering and leaving each node.
+ */
+class SourceCodeTraverser {
 	/**
-	 * Emits an event of entering AST node.
-	 * @param {ASTNode} node A node which was entered.
-	 * @returns {void}
+	 * The language of the source code being traversed.
+	 * @type {Language}
+	 */
+	#language;
+
+	/**
+	 * Map of languages to instances of this class.
+	 * @type {WeakMap<Language, SourceCodeTraverser>}
+	 */
+	static instances = new WeakMap();
+
+	/**
+	 * Creates a new instance.
+	 * @param {Language} language The language of the source code being traversed.
 	 */
-	enterNode(node) {
-		this.applySelectors(node, false);
-		this.currentAncestry.unshift(node);
+	constructor(language) {
+		this.#language = language;
+	}
+
+	static getInstance(language) {
+		if (!this.instances.has(language)) {
+			this.instances.set(language, new this(language));
+		}
+
+		return this.instances.get(language);
 	}
 
 	/**
-	 * Emits an event of leaving AST node.
-	 * @param {ASTNode} node A node which was left.
+	 * Traverses the given source code synchronously.
+	 * @param {SourceCode} sourceCode The source code to traverse.
+	 * @param {SafeEmitter} emitter The emitter to use for events.
 	 * @returns {void}
+	 * @throws {Error} If an error occurs during traversal.
 	 */
-	leaveNode(node) {
-		this.currentAncestry.shift();
-		this.applySelectors(node, true);
+	traverseSync(sourceCode, emitter) {
+		const esquery = new ESQueryHelper(emitter, {
+			visitorKeys: sourceCode.visitorKeys ?? this.#language.visitorKeys,
+			fallback: vk,
```

Comment:
```suggestion
			fallback: vk.getKeys,
```

`fallback` should be a function.